pkg/platform/kube/provisioner/apigateway/provisioner.go

```diff
@@ -0,0 +1,366 @@
+	ingresses := map[string]*ingress.Resources{}
+
+	// always try to remove previous canary ingress first, because
+	// nginx returns 503 on all requests if primary service == secondary service. (happens on every promotion)
+	// so during promotion all requests will be sent to the primary ingress
+	p.tryRemovePreviousCanaryIngress(ctx, apiGateway)
+
+	if len(upstreams) == 1 {
+
+		// create just a single ingress
+		ingressResources, err := p.generateNginxIngress(ctx, apiGateway, upstreams[0])
+		if err != nil {
+			return errors.Wrap(err, "Failed to generate nginx ingress")
+		}
+
+		ingresses[ingressResources.Ingress.Name] = ingressResources
+
+	} else if len(upstreams) == 2 {
+		var canaryUpstream platform.APIGatewayUpstreamSpec
+		var baseUpstream platform.APIGatewayUpstreamSpec
+
+		// determine which upstream is the canary one
+		if upstreams[0].Percentage != 0 {
+			baseUpstream = upstreams[1]
+			canaryUpstream = upstreams[0]
+		} else if upstreams[1].Percentage != 0 {
+			baseUpstream = upstreams[0]
+			canaryUpstream = upstreams[1]
+		} else {
+			return errors.New("Percentage must be set on one of the upstreams (canary)")
+		}
+
+		// validity check
+		if canaryUpstream.Percentage > 100 || canaryUpstream.Percentage < 1 {
+			return errors.New("The canary upstream percentage must be between 1 and 100")
+		}
+
+		// add the base ingress
+		baseIngressResources, err := p.generateNginxIngress(ctx, apiGateway, baseUpstream)
+		if err != nil {
+			return errors.Wrap(err, "Failed to generate the base nginx ingress")
+		}
+		ingresses[baseIngressResources.Ingress.Name] = baseIngressResources
+
+		// add the canary ingress
+		canaryIngressResources, err := p.generateNginxIngress(ctx, apiGateway, canaryUpstream)
+		if err != nil {
+			return errors.Wrap(err, "Failed to generate the canary nginx ingress")
+		}
+		ingresses[canaryIngressResources.Ingress.Name] = canaryIngressResources
+	}
+
+	// create ingresses
+	// must be done synchronously, first primary and then canary
+	// otherwise, when there is only canary ingress, the endpoint will not work (nginx behavior)
+	for ingressName, ingressResources := range ingresses {
+		if _, _, err := p.ingressManager.CreateOrUpdateResources(ingressResources); err != nil {
+			p.Logger.WarnWithCtx(ctx, "Failed to create/update api-gateway ingress resources",
+				"err", errors.Cause(err),
+				"ingressName", ingressName,
+				"appliedIngressNames", appliedIngressNames)
+			return errors.New("Failed to create/update api-gateway ingress resources")
+		}
+
+		appliedIngressNames = append(appliedIngressNames, ingressName)
+	}
+
+	// sleep 4 seconds as a safety, so nginx will finish updating the ingresses properly (it takes time)
+	p.Logger.DebugWithCtx(ctx,
+		"Updated nginx ingresses, sleeping for 4 seconds so nginx will stabilize",
+		"apiGatewayName", apiGateway.Name)
+	time.Sleep(4 * time.Second)
+
+	return nil
+}
+
+func (p *Provisioner) DeleteAPIGateway(ctx context.Context, namespace, name string) {
```

Comment:
Please see `functionres/lazy.go` and provide the same interface: List/Get/CreateOrUpdate/Delete for starters